src/connection/host_lookup.c

```diff
@@ -0,0 +1,435 @@
+  if (host_conf_Use == 0)
+    {
+      return gethostbyname (name);
+    }
+
+  hp = host_lookup_internal (name, NULL, HOSTNAME_TO_IPADDR);
+
+  if (hp == NULL)
+    {
+      return NULL;
+    }
+
+  return hp;
+}
+
+//gethostbyname_r_uhost (const char *hostname, struct hostent *out_hp)
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+int
+gethostbyname_r_uhost (const char *name,
+		       struct hostent *ret, char *buf, size_t buflen, struct hostent **result, int *h_errnop)
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+struct hostent *
+gethostbyname_r_uhost (const char *name, struct hostent *ret, char *buf, size_t buflen, int *h_errnop)
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+int
+gethostbyname_r_uhost (const char *name, struct hostent *ret, struct hostent_data *ht_data)
+#else
+struct hostent *
+gethostbyname (const char *name)
+#error "HAVE_GETHOSTBYNAME_R"
+#endif				/* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif				/* HAVE_GETHOSTBYNAME_R */
+{
+
+  struct hostent *hp_memcpy = NULL;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+//err also modified by callee
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+      return gethostbyname_r (name, ret, buf, buflen, result, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+      return gethostbyname_r (name, ret, buf, buflen, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+      return gethostbyname_r (name, ret, &ht_data);
+#else
+      return gethostbyname (name);
+#error "HAVE_GETHOSTBYNAME_R"
+#endif
+#endif /* !HAVE_GETHOSTBYNAME_R */
+    }
+
+  hp_memcpy = host_lookup_internal (name, NULL, HOSTNAME_TO_IPADDR);
+  memcpy ((void *) ret, (void *) hp_memcpy, sizeof (struct hostent));
+
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+  (*result) = (struct hostent *) malloc (sizeof (struct hostent));
+  memcpy ((void *) *result, (void *) hp_memcpy, sizeof (struct hostent));
+  return 0;
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+
+  return ret;
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+
+  return 0;
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif /* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif /* HAVE_GETHOSTBYNAME_R */
+}
+
+int
+getnameinfo_uhost (struct sockaddr *addr, socklen_t addrlen, char *host, size_t hostlen, char *serv, size_t servlen,
+		   int flags)
+{
+  struct hostent *hp;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return getnameinfo (addr, addrlen, host, hostlen, NULL, 0, NI_NOFQDN);
+    }
+
+  hp = host_lookup_internal (NULL, addr, IPADDR_TO_HOSTNAME);
+
+  if (hp == NULL)
+    {
+      return EAI_NODATA;
```

Comment:
EAI_NODATA Is there a reason to use this error code?
How about using the error code returned by the getnameinfo() function?
The getnameinfo () function returns the following error code.
EAI_AGAIN, EAI_BADFLAGS,  EAI_FAIL,  EAI_FAMILY, EAI_MEMORY, EAI_NONAME, EAI_OVERFLOW, EAI_SYSTEM

Reply:
Oh, this is my mistake. 

I was confused with getaddrinfo().
I will change it as **EAI_NONAME**